src/udp.rs

```diff
@@ -39,103 +114,293 @@ impl UdpSocket {
+        // this will wrap the pointer accordingly to the TX `free_size`.
+        // safe to cast to `u16` because the maximum buffer size in w5500 is 16 KB!
+        let new_write_pointer = write_pointer.wrapping_add(write_data.len() as u16);
+        self.socket.set_tx_write_pointer(bus, new_write_pointer)?;
+
+        self.block_send_command(bus)?;
+
+        Ok(write_data.len())
+    }
+
+    /// Sets the socket to [`socketnCommand::Send`] and block flushes the TX buffer
+    fn block_send_command<SpiBus: Bus>(
+        &self,
+        bus: &mut SpiBus,
     ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
-        // TODO increase longevity by cycling through buffer, instead of always writing to 0
-        // TODO ensure write is currently possible
-        self.socket
-            .set_tx_read_pointer(bus, 0)
-            .and_then(|_| bus.write_frame(self.socket.tx_buffer(), 0, send_buffer))
-            .and_then(|_| {
-                self.socket
-                    .set_tx_write_pointer(bus, send_buffer.len() as u16)
-            })
-            .and_then(|_| self.socket.command(bus, socketn::Command::Send))?;
+        // Send the data.
+        self.socket.command(bus, socketn::Command::Send)?;
 
         loop {
-            if self.socket.get_tx_read_pointer(bus)? == self.socket.get_tx_write_pointer(bus)? {
-                if self.socket.has_interrupt(bus, socketn::Interrupt::SendOk)? {
-                    self.socket
-                        .reset_interrupt(bus, socketn::Interrupt::SendOk)?;
-                    return Ok(());
-                } else if self
-                    .socket
-                    .has_interrupt(bus, socketn::Interrupt::Timeout)?
-                {
-                    self.socket
-                        .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
-                    return Err(NbError::Other(UdpSocketError::WriteTimeout));
-                }
+            match self.try_flush_tx(bus) {
+                Err(NbError::WouldBlock) => {}
+                result => return result,
             }
         }
     }
 
+    fn try_flush_tx<SpiBus: Bus>(
+        &self,
+        bus: &mut SpiBus,
+    ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
+        let tx_read = self.socket.get_tx_read_pointer(bus)?;
+        let tx_write = self.socket.get_tx_write_pointer(bus)?;
+
+        if tx_read == tx_write {
+            if self.socket.has_interrupt(bus, socketn::Interrupt::SendOk)? {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::SendOk)?;
+
+                return Ok(());
+            } else if self
+                .socket
+                .has_interrupt(bus, socketn::Interrupt::Timeout)?
+            {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
+
+                return Err(NbError::Other(UdpSocketError::WriteTimeout));
+            }
```

Comment:
Cascading the ifs is unnecessary because the previous if always returns.

```suggestion
            }
            
            if self
                .socket
                .has_interrupt(bus, socketn::Interrupt::Timeout)?
            {
                self.socket
                    .reset_interrupt(bus, socketn::Interrupt::Timeout)?;

                return Err(NbError::Other(UdpSocketError::WriteTimeout));
            }
```